### PR TITLE
Create facilities tab & display text search results

### PR DESCRIPTION
**Pull request by kellyi**
## Overview

This PR adds a facilities list tab which will display a list of text search results as depicted below when the facilities results return from a search of the API. I also added a free text search which currently searches on the `name` `address` and `country_name` fields to narrow down the list of visible results, as well as a "Download CSV" functionality which downloads a list of CSV items.

Additionally, this PR also:

- replaces papaparse with a custom utility function to turn a 2d array into a string, which circumvents a bug whereby JSON.parse would sometimes try to parse the OAR ID as a number but fail and instead print `NaN` to the CSV
- adds tests for that newly added utility function
- adjusts the contributor list to list them in alphabetical order by name
- adjusts the app's DOM structure to make it possible to scroll subcomponents rather than scrolling the entire app
- changes the app name in `package.json` from "redpanda" (which was the legacy name?) to "open apparel registry"

Connects #209
Connects #227
Connects #127 
Connects #123 

## Demo

facilities-list

CSV: 
[facilities.csv.txt](https://github.com/open-apparel-registry/open-apparel-registry/files/2927348/facilities.csv.txt)

No results for a search:

screen shot 2019-03-04 at 11 47 10 am

## Notes

I made #244 for the map interactions because it will need some dedicated thought about how to handle the various possible kinds of interaction.

## Testing

- get this branch then `./scripts/update`
- run `./scripts/test` and verify that everything passes
- `./scripts/resetdb`
- visit the main map page and
    - verify that the contributors are now listed in alphabetical order by name
    - make a search and verify that you are navigated to the facilities tab on success
    - verify that the free text search works as expected
    - verify that you can download a list of facilities, filtered by your free text search
    - verify that you can scroll through the facilities and that clicking on any of the list items takes you to its details route
- on the Macbook-sized viewport, check that the following pages scroll correctly following the changes I made to the DOM:
   - `/auth/register`
   - `/contribute`
   - `/profile/:id/`
   - `/lists`
   - `/lists/:id`
- download a facility list and verify that that still works, following the changes to replace papaparse

### CODE REVIEW

**Comment by jwalgran** [comment]
Looking at this now

**Comment by jwalgran** [comment]
I'm seeing an odd click target bug when switching tabs in Chrome. Only clicks at the very top or bottom edges of the tab are triggering the state change. Clicks in the middle of the tab are triggering the Material UI animation, but not a state change.

**Comment by kellyi** [comment]
> I'm seeing an odd click target bug when switching tabs in Chrome.

Weird! I wonder if I set up the click handlers incorrectly. I tested staging and it's not happening there.

**Comment by kellyi** [comment]
Interesting, I must need to add an ID to another element:

screen shot 2019-03-04 at 3 04 11 pm

**Comment by kellyi** [comment]
Turns out the correct function signature is:

```
  handleChange = (event, value) => {
    this.setState({ value });
  };
```

Adjusted it in 1a05babbb476104b76dd6d5244243a6086a507a1

**Comment by jwalgran** [comment]
It is a bit frustrating to loose scroll state on the search results. When navigating back from the details view. Have we done any scroll restoration on other projects that could possibly apply here? This react-router document talks about the issue but does not include a general solution for our issue. https://github.com/ReactTraining/react-router/blob/master/packages/react-router-dom/docs/guides/scroll-restoration.md

**Comment by kellyi** [comment]
> It is a bit frustrating to loose scroll state on the search results. When navigating back from the details view. Have we done any scroll restoration on other projects that could possibly apply here?

Make sense. I haven't ever worked on it; the nearest thing to this is the parcel cards sidebar in GARP for which we declined to try to set up scroll restoration.

I'm happy to look into it, but I wonder if it'd make sense as a new card?

**Comment by jwalgran** [comment]
>I'm happy to look into it, but I wonder if it'd make sense as a new card?
Yes. We can break scroll state restoration into a separate issue.

File: src/app/src/components/FilterSidebarFacilitiesTab.jsx [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+    sortFacilitiesAlphabeticallyByName,
+} from '../util/util';
+
+import COLOURS from '../util/COLOURS';
+
+import { filterSidebarStyles } from '../util/styles';
+
+const SEARCH_TERM_INPUT = 'SEARCH_TERM_INPUT';
+
+const makeFacilityListHeightPercentageString = () => {
```
Comment (jwalgran) [inline src/app/src/components/FilterSidebarFacilitiesTab.jsx]: On my 1680x1050 screen there is some whitespace between the scrolling results list and the footer. If I shrink the browser window height the whitespace goes away. I looks like the conditional styling set up by this function may be the reason. Is there a way we can make sure that the scrolling area always fills the container?

screen shot 2019-03-04 at 1 29 44 pm

**Comment by kellyi** [comment]
> If you run a search, click back to the search tab, change one of the filters, then click back to the "Facilities" tab you are now in a kind of inconsistent state. We may want to consider one of the following changes to make sure the state of the two tabs are always consistent.

How would you feel about dumping any existing facilities data altogether when the state of the filters changes? That'd also solve an inconsistency between prior data and the new filters before "search" has been clicked.

**Comment by jwalgran** [comment]
>How would you feel about dumping any existing facilities data altogether when the state of the filters changes?
I think that is a good idea in general, but what would the facilities tab say at that point? If it reverts to the "No facilities were found for that search" message we are in the same situation.

File: src/app/src/__tests__/utils.tests.js [hunk 2]
```diff
@@ -716,3 +720,135 @@ it('gets a `token` from a querystring', () => {
...
+            3,
+            4,
+        ],
+    ];
+    const expectedNumericArrayMatch = '1,2\n3,4\n';
+    expect(joinDataIntoCSVString(numericArray)).toBe(expectedNumericArrayMatch);
+
+    const stringArray = [
+        [
+            'hello',
```
[the remark this answers is not in the document]
Comment (kellyi) [reply src/app/src/__tests__/utils.tests.js]: Good idea! I added the problematic one in 5c028626a3475f63a5a76ee31f6b5177a67ec8d9

File: src/app/src/components/FilterSidebarFacilitiesTab.jsx [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+                    >
+                        Download CSV
+                    </Button>
+                </div>
+            </Typography>
+            <div style={facilitiesTabStyles.listHeaderTextSearchStyles}>
+                <label
+                    htmlFor={SEARCH_TERM_INPUT}
+                >
+                    Search results
```
[the remark this answers is not in the document]
Comment (kellyi) [reply src/app/src/components/FilterSidebarFacilitiesTab.jsx]: Sounds good. 42bbd938dbbbac20b9750d62e6928da1ea9dab33 renames these from using `searchTerm` to using `filterText`.

**Comment by kellyi** [comment]
> I think that is a good idea in general, but what would the facilities tab say at that point? If it reverts to the "No facilities were found for that search" message we are in the same situation.

Interesting. Maybe we just don't show the tab at all if there are no facilities in the state to display? I'll try it out locally to see how that works.

**Comment by kellyi** [comment]
9a189ee1b245bfed700608a67cc14830bb1ee655 creates a new "no facilities to display" state for the sidebar tab so that it will look like this when there's nothing to show:

screen shot 2019-03-05 at 11 00 04 am

In addition, I updated the application state to clear facilities data when any of the search filters change and to drop the markers from the map when the facilities data is cleared. For the latter, I kept the layer around, but just replace the data with an empty geojson feature collection so that no markers will show.

**Comment by kellyi** [comment]
cd088df48235d13d08749e6f617e1b88b675dc1e updates the facilities list to have its height set dynamically based on the viewport size

@jwalgran this is ready for another look!

**Comment by kellyi** [comment]
https://github.com/open-apparel-registry/open-apparel-registry/issues/250 is for scroll restoration

It occurred to me that we may be able to find an infinite scroll library that supports scroll restoration innately. That would also solve potential performance problems if the list of search results was in the thousands.

File: src/app/package.json [hunk 4]
```diff
@@ -40,8 +38,7 @@
         "build": "react-scripts build",
         "test": "CI=1 react-scripts test",
         "lint": "eslint src/ --ext .js --ext .jsx",
-        "eject": "react-scripts eject",
-        "deploy": "npm run build && firebase deploy"
```
[the remark this answers is not in the document]
Comment (kellyi) [reply src/app/package.json]: Just a note: I have not changed this yet but I will update the commit message when squashing and rebasing before I merge.

**Comment by kellyi** [comment]
Rebased this on develop & squashed.

File: src/app/package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (kellyi) [reply src/app/package.json]: Updated commit is 
6fd1a9b

**Comment by jwalgran** [comment]
Looking at this now

File: src/app/src/components/FilterSidebarFacilitiesTab.jsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (jwalgran) [reply src/app/src/components/FilterSidebarFacilitiesTab.jsx]: I am still seeing "Search" on the "Facilities" tab. Consider changing the label to "Filter results" and the placeholder text to "Filter by name, address, or country"

screen shot 2019-03-05 at 2 04 46 pm

**Comment by kellyi** [comment]
> I am still seeing "Search" on the "Facilities" tab. Consider changing the label to "Filter results" and the placeholder text to "Filter by name, address, or country"

Oof, must have forgot to update this. Will make this change before merging.

File: src/app/src/components/FilterSidebarFacilitiesTab.jsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (kellyi) [reply src/app/src/components/FilterSidebarFacilitiesTab.jsx]: I replaced the "Search" text with "Filter" in the adjusted last commit here.

**Comment by kellyi** [comment]
Thanks for your help with this!

**Pull request merged**